Source/Core/DolphinQt2/MenuBar.cpp

```diff
@@ -271,3 +309,58 @@ void MenuBar::InstallWAD()
 
   result_dialog.exec();
 }
+
+void MenuBar::PerformOnlineUpdate(const std::string& region)
+{
+  const int confirm = QMessageBox::question(
+      nullptr, tr("Confirm"), tr("Connect to the Internet and perform an online system update?"));
+  if (confirm != QMessageBox::Yes)
+    return;
+
+  QProgressDialog dialog;
+  dialog.setCancelButton(nullptr);
+  dialog.setLabelText(tr("Preparing to update...\nThis can take a while. Do not close Dolphin."));
+  dialog.setWindowTitle(tr("Updating"));
+  dialog.setWindowModality(Qt::WindowModal);
+  dialog.setWindowFlags(dialog.windowFlags() & ~Qt::WindowCloseButtonHint);
+  dialog.setMinimumSize(360, 120);
+
+  std::future<WiiUtils::UpdateResult> result = std::async(
+      std::launch::async, &WiiUtils::DoOnlineUpdate,
+      [&dialog](size_t processed, size_t total, u64 title_id) {
+        Core::QueueHostJob(
+            [&dialog, processed, total, title_id]() {
+              dialog.setRange(0, static_cast<int>(total));
+              dialog.setValue(static_cast<int>(processed));
+              dialog.setLabelText(
+                  tr("Updating title %1...\nThis can take a while. Do not close Dolphin.")
+                      .arg(title_id, 16, 16, QLatin1Char('0')));
+            },
+            true);
+      },
+      region);
+
+  dialog.exec();
+  while (result.wait_for(std::chrono::milliseconds(50)) != std::future_status::ready)
```

Comment:
I see. I used exec() when I was trying QFuture, QFutureWatcher and QtConcurrent (which unfortunately doesn't allow proper progress reporting). And I guess exec() will let Qt run the event loop, so doing it manually should be unnecessary.

I will add a finish/error callback, which should help with better error reporting.